### PR TITLE
drivers: sensor: add driver and example for DPS310

### PR DESCRIPTION
**Pull request by creiter32**
Add driver and example for Infineon DPS310 temperature and pressure sensor.

This is a rebased version of #20187, since that PR went stale.

Signed-off-by: Christoph Reiter christoph.reiter@infineon.com

### CODE REVIEW

**Review by oyvindronningstad — APPROVED** [review]
Looked quickly through. Just a few nits, this looks ready to me.

File: samples/sensor/dps310/CMakeLists.txt
Comment (galak) [inline samples/sensor/dps310/CMakeLists.txt]: You shouldn't need to set DTC_OVERLAY_FILE.  You can use `app.overlay` for the dts file name and the build system will pick it up for you.  (see samples/sensor/amg88xx for an example).

File: samples/sensor/dps310/CMakeLists.txt
Comment (galak) [inline samples/sensor/dps310/CMakeLists.txt]: This should now be `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`.  Recent changed merged in upstream Zephyr.

**Review by galak — DISMISSED** [review]
I converted the DT bits to the new DT_INST macro's in include/devicetree.h.  Made a few other comments.

**Comment by creiter32** [comment]
Applied the suggested fixes and rebased on latest master.

**Comment by creiter32** [comment]
@galak Sorry, I seem to have accidentally undone your changes, can you please re-apply them again?

**Comment by galak** [comment]
> @galak Sorry, I seem to have accidentally undone your changes, can you please re-apply them again?

will do.

**Comment by galak** [comment]
Redid my changes.  You might want to think about moving `DPS310 temperature oversampling` and `DPS310 pressure oversampling` from Kconfig into dts properties.  Since you support more than one instance of the sensor, I wonder if a user will always want those settings the same for both instances.

**Comment by creiter32** [comment]
Rebased to fix conflicts.

> Redid my changes. You might want to think about moving DPS310 temperature oversampling and DPS310 pressure oversampling from Kconfig into dts properties. Since you support more than one instance of the sensor, I wonder if a user will always want those settings the same for both instances.

If this is not a hard requirement, I would leave it as is, because the sensor is an environmental sensor and it there will be very few people actually using more than one instance, let alone with different settings.

**Comment by creiter32** [comment]
@galak Is there anything else I need to do to get this merged?

File: drivers/sensor/dps310/dps310.c
Comment (MaureenHelm) [inline drivers/sensor/dps310/dps310.c:469]: use `sys_le24_to_cpu` instead

File: tests/drivers/build_all/i2c.dtsi
Comment (MaureenHelm) [inline tests/drivers/build_all/i2c.dtsi:525]: move this change to the previous commit, so it goes with `tests/drivers/build_all/sensors_a_h.conf`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/dps310/dps310.c
Comment (creiter32) [reply drivers/sensor/dps310/dps310.c:463]: Replaced with the `sys_get_be*` functions instead, because they fit better, additionally the comment about the values being little endian was wrong, they are big endian.

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@galak please revisit

**Comment by creiter32** [comment]
@galak any update?

**Comment by nandojve** [comment]
Hi all, Will this be available for 2.3? It has been a long journey and is almost there.
@creiter32 just curious, what shield do you recommend to test this Sensor?

**Comment by creiter32** [comment]
@nandojve For development I used this one: https://www.infineon.com/cms/en/product/evaluation-boards/s2go-pressure-dps310/

**Comment by carlescufi** [comment]
@galak please take another look, I've been asked to consider this for 2.3

**Comment by carlescufi** [comment]
Do not merge yet until #25214 has been merged.

**Comment by galak** [comment]
We need to update `DT_HAS_NODE_STATUS_OKAY` to DT_NODE_HAS_STATUS(..., okay)

File: drivers/sensor/dps310/dps310.c
Comment (galak) [inline drivers/sensor/dps310/dps310.c:3]: Remove this, duplicated on line 11.

**Comment by carlescufi** [comment]
@creiter32 would you mind applying the changes described above by @galak so we can merge this in time for 2.3?

**Comment by galak** [comment]
Here's my diff for the change.:

```
diff --git a/drivers/sensor/dps310/dps310.c b/drivers/sensor/dps310/dps310.c
index d3a363db26..65c6d3a344 100644
--- a/drivers/sensor/dps310/dps310.c
+++ b/drivers/sensor/dps310/dps310.c
@@ -1,7 +1,5 @@
 /* Driver for Infineon DPS310 temperature and pressure sensor */
 
-#define DT_DRV_COMPAT infineon_dps310
-
 /*
  * Copyright (c) 2019 Infineon Technologies AG
  *
@@ -725,7 +723,7 @@ static const struct sensor_driver_api dps310_api_funcs = {
 
 /* support up to 2 instances */
 
-#if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(0))
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
 static struct dps310_data dps310_data_0;
 static const struct dps310_cfg dps310_cfg_0 = {
        .i2c_bus_name = DT_INST_BUS_LABEL(0),
@@ -737,7 +735,7 @@ DEVICE_AND_API_INIT(dps310, DT_INST_LABEL(0), dps310_init,
                    CONFIG_SENSOR_INIT_PRIORITY, &dps310_api_funcs);
 #endif
 
-#if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(1))
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
 static struct dps310_data dps310_data_1;
 static const struct dps310_cfg dps310_cfg_1 = {
        .i2c_bus_name = DT_INST_BUS_LABEL(1),
```

**Comment by creiter32** [comment]
@galak @carlescufi Applied the requested changes and rebased. Allow edits by maintainers
 is also enabled in case other small fixes are needed.

**Comment by carlescufi** [comment]
> @galak @carlescufi Applied the requested changes and rebased. Allow edits by maintainers
> is also enabled in case other small fixes are needed.

Thank you @creiter32. This is merged now.

**Pull request merged**
